8346094: Harden X509CertImpl.getExtensionValue for NPE cases

ksreenath-ora opened the pull request:
Earlier code will trigger NPE if the certificate does not contain the extensions or if the requested extensions does not exist. The better approach for hardening **getExtensionValue** here is to to check for NULL explicitly before calling **getExtensionValue()** and avoding try-catch block which ensures the readability and maintainability.

After scanning in multiple places where invokng getExtensions on the X509CertInfo reference, the check for NULL is added in the **getKeyUsage()** as well while calling before **getExtensionValue()**

The associated tests are written and added in test class **CertificateExtensions**. Which will ensure to validate the 
**getExtensionValue()** and **getKeyUsage()** methods in **X509CertImpl** class.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8346094](https://bugs.openjdk.org/browse/JDK-8346094): Harden X509CertImpl.getExtensionValue for NPE cases (**Bug** - P4)


### Reviewers
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)
 * [Weijun Wang](https://openjdk.org/census#weijun) (@wangweij - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23315/head:pull/23315` \
`$ git checkout pull/23315`

Update a local copy of the PR: \
`$ git checkout pull/23315` \
`$ git pull https://git.openjdk.org/jdk.git pull/23315/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23315`

View PR using the GUI difftool: \
`$ git pr show -t 23315`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23315.diff">https://git.openjdk.org/jdk/pull/23315.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23315#issuecomment-2638911491)
</details>

mpdonova commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,202 @@
+
```
Needs a copyright statement

ksreenath-ora replied:
added the copyright statement

mpdonova commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,202 @@
+         * else it is incorrect
+         */
+        try {
+            System.out.println("------ cert x509Certimpl ------ " + x509Certimpl);
```
The output might be a little easier to read if you put the "---- cert X509Certimpl ----" on its own line.

ksreenath-ora replied:
was added these during the debug and continued the same in draft PR. I have removed these statements

mpdonova commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,202 @@
+         * Certificate is created without extensions. Invoking getExtensionValue with oid must return NULL
+         * else it is incorrect
+         */
+        try {
```
These try/catch blocks around the assert statements seem redundant. The assert methods already throw a RuntimeException with a message.

ksreenath-ora replied:
Addressed and consolidated all try-catch blocks

mpdonova replied:
Is this try/catch necessary at all? The methods in Asserts throw RuntimeExceptions already and the "test setup" code like createCertificateExceptions() throws other exceptions which is fine because main() is declared to throw Exception.

nibjen commented on src/java.base/share/classes/sun/security/x509/X509CertImpl.java:
The copyright year should be 2025

nibjen commented on src/java.base/share/classes/sun/security/x509/X509CertImpl.java:
Copyright year should be 2025

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
Line indentation issue

nibjen commented on src/java.base/share/classes/sun/security/x509/X509CertImpl.java:
Copyright year should be 2025 since the file is changed

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+/**
+ * @test
+ * @bug 8346094
+ * @summary validating getExtensionValue and getKeyUsage with specified and unspecified extensions on the X509Certificate.
```
Check line indentation issue

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+ * @summary validating getExtensionValue and getKeyUsage with specified and unspecified extensions on the X509Certificate.
+ * @library /test/lib
+ * @modules java.base/sun.security.x509
+ * java.base/sun.security.util
```
Check line indentation issue

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+             * Certificate is created with extensions. Invoking getExtensionValue with oid must not return NULL
+             * else it is incorrect
+             */
+            x509Certimpl.getInfo().setExtensions(createCertificateExtensions(x509Certimpl.getInfo().getKey().getKey()));
```
Line indentation issue

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+            x509Certimpl.getInfo().setExtensions(createCertificateExtensions(x509Certimpl.getInfo().getKey().getKey()));
+            Asserts.assertNotNull(x509Certimpl.getExtensionValue("2.5.29.17"));
+            /**
+             * Certificate is created with extensions. Invoking getExtensionValue with invalid oid must return NULL
```
Same comment as above

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+            x509CertInfo.setIssuer(issuer);
+            x509CertInfo.setSubject(subject);
+
+            //create and set the subject and issuer unique identity
```
Give space after //

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+            byte[] signedData = signature.sign();
+            byte[] signedCert = {};
+
+            x509Certimpl = new X509CertImpl(x509CertInfo, AlgorithmId.get("SHA256withRSA"), signedData, signedCert);
```
Line indentation

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+        return x509Certimpl;
+    }
+
+    public static sun.security.x509.CertificateExtensions createCertificateExtensions(PublicKey publicKey) throws IOException, NoSuchAlgorithmException {
```
Same comment as above

nibjen commented on test/jdk/java/security/cert/X509Certificate/CertificateExtensions.java:
```diff
@@ -0,0 +1,199 @@
+
+    public static sun.security.x509.CertificateExtensions createCertificateExtensions(PublicKey publicKey) throws IOException, NoSuchAlgorithmException {
+        // Create Extensions
+        sun.security.x509.CertificateExtensions certificateExtensions = new sun.security.x509.CertificateExtensions();
```
Same comment

ksreenath-ora replied:
addressed indentation

seanjmullan commented:
/reviewers 2

seanjmullan commented:
Please don't integrate until someone from the Security Group reviews this PR. Thanks.

seanjmullan commented:
@jnimeh Can you please review this PR?

wangweij commented:
I'm wondering how necessary this fix is. These are internal classes, only called inside JDK, where some pre-conditions are always met. Unless someone explicitly calls `x509Certimpl.getInfo().setExtensions(null)` (as done in the test), it seems like both the `info` and `extensions` fields should never be null.

If you’re concerned about misuse of these methods leading to potential future issues, consider adding comments to clarify their expected usage. You could also include `assert` statements or `Objects.requireNonNull` calls to enforce these preconditions.

ksreenath-ora commented:
> > I'm wondering how necessary this fix is. These are internal classes, only called inside JDK, where some pre-conditions are always met. Unless someone explicitly calls `x509Certimpl.getInfo().setExtensions(null)` (as done in the test), it seems like both the `info` and `extensions` fields should never be null.
> > If you’re concerned about misuse of these methods leading to potential future issues, consider adding comments to clarify their expected usage. You could also include `assert` statements or `Objects.requireNonNull` calls to enforce these preconditions.
> 
> I agree with Weijun. The `info` field should never be `null` under normal circumstances. Checking if it is `null` would only make sense if there is a bug somewhere else in the code, and in that case, the bug should be fixed.

I made these changes for the following reasons:

Null Check on info: The X509CertImpl ref can be created with a null value for info, so I added a null check to prevent potential issues.
Consistency Across Internal Classes: Other internal classes already perform the same null check on info. To maintain consistency, I added this change here as well.
Null Check on extensions: The info object can be created without extensions. Since other parts of the code already handle potential null values for extensions, I added this check to maintain uniformity.

wangweij commented:
I take back my word about `extensions` -- it can be null. However, `X509CertImpl` should never have a null `info`. I’d prefer to add `Objects.requireNonNull` in its constructor. Your approach of returning null instead of throwing an exception when `info == null` actually hides this problem. If there are other places when `info == null` is checked, please remove them.

ksreenath-ora commented:
> I take back my word about `extensions` -- it can be null. However, `X509CertImpl` should never have a null `info`. I’d prefer to add `Objects.requireNonNull` in its constructor. Your approach of returning null instead of throwing an exception when `info == null` actually hides this problem. If there are other places when `info == null` is checked, please remove them.

Yes, as you suggested, this approach to the code change is much clearer. I will remove the null check on info and replace it with Objects.requireNonNull in the constructor of the X509CertImpl class. This will result in modifications to the test class to accommodate the new implementation. Additionally, I will move the CertificateExtensions class to the internal package to facilitate testing of the internal methods."

wangweij commented on test/jdk/sun/security/x509/X509CertImpl/CertificateExtensions.java:
Another test `V3Certificate` in the same directory references the internal `CertificateExtensions` class by its simple name. There is a conflict with this new test and it does not compile. Consider renaming your new test. Thanks.

wangweij commented on test/jdk/sun/security/x509/X509CertImpl/CertExtensions.java:
```diff
@@ -27,8 +27,6 @@
  *          unspecified extensions on the X509Certificate.
  * @library /test/lib
- * @modules java.base/sun.security.x509
- *          java.base/sun.security.util
```
Why remove these `@modules` lines? They are still necessary to access internal classes.

wangweij commented:
There is still one `if (info != null)` check in `getTBSCertificate()`.

Also, you can combine
```
        Objects.requireNonNull(info);
        this.info = info;
```
to
```
        this.info = Objects.requireNonNull(info);
```.

ksreenath-ora commented:
> There is still one `if (info != null)` check in `getTBSCertificate()`.
> 
> Also, you can combine
> 
> ```
>         Objects.requireNonNull(info);
>         this.info = info;
> ```
> 
> to
> 
> ```
>         this.info = Objects.requireNonNull(info);
> ```

addresses @wangweij

wangweij commented:
Only a small nit. Copyright year for `TestHostnameChecker.java` should also be updated.

ksreenath-ora commented:
/integrate

coffeys reviewed (approved):
LGTM

ksreenath-ora commented:
/integrate

wangweij commented:
/sponsor
